plugins/modem-manager/fu-dfota-updater.c

```diff
@@ -0,0 +1,382 @@
+	acks_expected = chunk_size / 1024;
+	/* pad every chunk to 2048 bytes to received correct amount of ACKs */
+	chunk_bytes = fu_bytes_pad(fu_chunk_get_bytes(chk), 0x800);
+
+	if (!fu_io_channel_write_bytes(self->io_channel,
+				       chunk_bytes,
+				       1500,
+				       FU_IO_CHANNEL_FLAG_NONE,
+				       error)) {
+		g_prefix_error(error, "failed to upload firmware to the device: ");
+		return FALSE;
+	}
+	if (acks_expected == 0)
+		return TRUE;
+
+	ack_bytes = fu_io_channel_read_bytes(self->io_channel,
+					     acks_expected,
+					     FU_DFOTA_UPDATER_TIMEOUT_SECS * 1000,
+					     FU_IO_CHANNEL_FLAG_NONE,
+					     error);
+
+	if (ack_bytes == NULL) {
+		g_prefix_error(error, "failed to read response: ");
+		return FALSE;
+	}
+
+	ack_result = g_bytes_get_data(ack_bytes, &ack_size);
+	if (ack_size != acks_expected) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "expected %" G_GSIZE_FORMAT " ACKs, got %" G_GSIZE_FORMAT,
+			    acks_expected,
+			    ack_size);
+		return FALSE;
+	}
+	if (!g_regex_match(ack_regex, ack_result, 0, NULL)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "expected ACKs (A), got %s",
+			    ack_result);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dfota_updater_parse_upload_result(FuDfotaUpdater *self,
+				     gchar **checksum,
+				     gsize *size,
+				     GError **error)
+{
+	g_autoptr(GBytes) result_bytes = NULL;
+	g_autoptr(GRegex) result_regex = NULL;
+	g_autoptr(GMatchInfo) match_info = NULL;
+	g_autofree gchar *size_match = NULL;
+	g_autofree gchar *checksum_match = NULL;
+	const gchar *result = NULL;
+
+	/* +QFUPL: <filesize>,<hex checksum> */
+	result_regex = g_regex_new("\\r\\n\\+QFUPL:\\s*(\\d+),([0-9a-f]+)\\r\\n", 0, 0, NULL);
+
+	result_bytes = fu_io_channel_read_bytes(self->io_channel,
+						-1,
+						FU_DFOTA_UPDATER_TIMEOUT_SECS * 1000,
+						FU_IO_CHANNEL_FLAG_SINGLE_SHOT,
+						error);
+	if (result_bytes == NULL) {
+		g_prefix_error(error, "failed to read AT+QFUPL response: ");
+		return FALSE;
+	}
+	result = g_bytes_get_data(result_bytes, NULL);
+
+	if (g_strrstr(result, "\r\nOK\r\n") == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "upload command exited with error");
+		return FALSE;
+	}
+	if (!g_regex_match(result_regex, result, 0, &match_info)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "could not match QFUPL response");
+		return FALSE;
+	}
+	if (!g_match_info_matches(match_info)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "could not match size and checksum");
+		return FALSE;
+	}
+
+	size_match = g_match_info_fetch(match_info, 1);
+	checksum_match = g_match_info_fetch(match_info, 2);
+
+	g_debug("parsed checksum '%s' and size '%s'", checksum_match, size_match);
+
+	*checksum = g_steal_pointer(&checksum_match);
```

Comment:
shouldn't this only be updated after the last `g_ascii_string_to_unsigned`?

Reply:
NVM I see the problem here.